File: flutter/lib/features/kyc/data/kyc_fields.dart
// PR-KYC-CORRECTION (2026-09-06): canonical KYC correction-field taxonomy
// — single source of truth for the rider app.
//
// Values mirror the admin checklist
// (web/src/components/admin/screens/kyc-management/correction-fields.ts)
// and the `kyc_editable_fields_allowlist` DB constraint. The ORDER below
// matches the onboarding form, so "first flagged field" routes the rider
// to the right step:
//   step 1 — Personal details   (fullName … currentAddress)
//   step 2 — Identity & bank    (aadhaarFront … ifscCode)
//   step 3 — Selfie & signature (profilePhoto, signature)

/// Canonical field keys in onboarding-form order.
const List<String> kycCorrectionFieldOrder = <String>[
  'fullName',
  'fatherName',
  'motherName',
  'dob',
  'email',
  'currentAddress',
  'aadhaarFront',
  'aadhaarBack',
  'panCard',
  'bankName',
  'accountNumber',
  'ifscCode',
  'profilePhoto',
  'signature',
];

/// Legacy / alternate server keys mapped onto the canonical ones
/// (the server's EditableField enum historically used 'name' and
/// 'address'; 'pan'/'selfie' are defensive aliases).
const Map<String, String> _kycFieldAliases = <String, String>{
  'name': 'fullName',
  'address': 'currentAddress',
  'pan': 'panCard',
  'selfie': 'profilePhoto',
};

/// Maps a raw flagged-field key onto its canonical form key
/// (aliases resolved; unknown keys pass through unchanged).
String normalizeKycFieldKey(String raw) {
  final trimmed = raw.trim();
  return _kycFieldAliases[trimmed] ?? trimmed;
}

/// Normalizes a raw flagged-field list: aliases resolved, unknown keys
/// dropped, de-duplicated, and ordered per the onboarding form.
List<String> normalizeKycEditableFields(List<String>? raw) {
  if (raw == null || raw.isEmpty) return const <String>[];
  final seen = raw
      .map((f) => _kycFieldAliases[f.trim()] ?? f.trim())
      .where(kycCorrectionFieldOrder.contains)
      .toSet();
  return kycCorrectionFieldOrder.where(seen.contains).toList();
}

/// The onboarding step (1-3) that owns the FIRST flagged field — the
/// step the "Correct the details" button deep-links to. Defaults to 1
/// when nothing is flagged.
int firstFlaggedKycStep(List<String>? raw) {
  final fields = normalizeKycEditableFields(raw);
  if (fields.isEmpty) return 1;
  final first = fields.first;
  if (first == 'profilePhoto' || first == 'signature') return 3;
  if (kycCorrectionFieldOrder.indexOf(first) >=
      kycCorrectionFieldOrder.indexOf('aadhaarFront')) {
    return 2;
  }
  return 1;
}

/// English labels for the flagged-field chips. Localized labels are
/// resolved through AppLocalizations by the caller where available;
/// these are the fallbacks.
const Map<String, String> kycCorrectionFieldLabels = <String, String>{
  'fullName': 'Full name',
  'name': 'Full name',
  'fatherName': "Father's name",
  'motherName': "Mother's name",
  'dob': 'Date of birth',
  'email': 'Email',
  'currentAddress': 'Address',
  'address': 'Address',
  'aadhaarFront': 'Aadhaar card (front)',
  'aadhaarBack': 'Aadhaar card (back)',
  'panCard': 'PAN card',
  'pan': 'PAN card',
  'bankName': 'Bank details',
  'accountNumber': 'Bank details',
  'ifscCode': 'Bank details',
  'profilePhoto': 'Profile photo',
  'selfie': 'Profile photo',
  'signature': 'Signature',
};

/// Label for a raw (un-normalized) key, with fallback to the key itself.
String kycCorrectionFieldLabel(String rawKey) =>
    kycCorrectionFieldLabels[rawKey] ?? kycCorrectionFieldLabels[_kycFieldAliases[rawKey]] ?? rawKey;

File: web/src/lib/kyc-fields.ts
/**
 * KYC-CORRECTION-P0-2026-09-08 (P0-1): canonical KYC correction-field
 * taxonomy — single source of truth for the admin panel.
 *
 * Mirrors `flutter/lib/features/kyc/data/kyc_fields.dart`. The order
 * matches the onboarding form (step 1: personal details, step 2:
 * identity & bank, step 3: selfie & signature) so the rider app's
 * `firstFlaggedKycStep` aligns with the admin's field picker.
 *
 * Used by:
 *  - `useKyc.ts` (KYC admin hook) — populates the editableFields array
 *    sent to the server when an admin clicks Reject / Request
 *    Correction
 *  - `KycDialogs.tsx` (admin reject dialog) — renders the field
 *    picker checkboxes
 *  - `rider.use-cases.ts` (server) — the editableFields allowlist
 *    gates which fields a rider can resubmit
 */

export type KycCorrectionField =
  | 'fullName'
  | 'fatherName'
  | 'motherName'
  | 'dob'
  | 'email'
  | 'currentAddress'
  | 'aadhaarFront'
  | 'aadhaarBack'
  | 'panCard'
  | 'bankName'
  | 'accountNumber'
  | 'ifscCode'
  | 'profilePhoto'
  | 'signature';

/** Canonical field keys in onboarding-form order. */
export const KYC_CORRECTION_FIELDS: ReadonlyArray<{
  key: KycCorrectionField;
  label: string;
  group: 'personal' | 'identity' | 'media';
}> = [
  { key: 'fullName', label: 'Full name', group: 'personal' },
  { key: 'fatherName', label: "Father's name", group: 'personal' },
  { key: 'motherName', label: "Mother's name", group: 'personal' },
  { key: 'dob', label: 'Date of birth', group: 'personal' },
  { key: 'email', label: 'Email', group: 'personal' },
  { key: 'currentAddress', label: 'Current address', group: 'personal' },
  { key: 'aadhaarFront', label: 'Aadhaar (front)', group: 'identity' },
  { key: 'aadhaarBack', label: 'Aadhaar (back)', group: 'identity' },
  { key: 'panCard', label: 'PAN card', group: 'identity' },
  { key: 'bankName', label: 'Bank name', group: 'identity' },
  { key: 'accountNumber', label: 'Account number', group: 'identity' },
  { key: 'ifscCode', label: 'IFSC code', group: 'identity' },
  { key: 'profilePhoto', label: 'Profile photo', group: 'media' },
  { key: 'signature', label: 'Signature', group: 'media' },
] as const;

/** All valid editable-field keys (server-side allowlist reference). */
export const KYC_CORRECTION_FIELD_KEYS: ReadonlyArray<KycCorrectionField> =
  KYC_CORRECTION_FIELDS.map((f) => f.key);

/** Human-friendly label for a field key. Falls back to the raw key. */
export function kycFieldLabel(key: string): string {
  const found = KYC_CORRECTION_FIELDS.find((f) => f.key === key);
  return found?.label ?? key;
}
